src/utilities/codeBlockHelpers.cy.ts

```diff
@@ -0,0 +1,229 @@
+  ],
+  "markup": "<div class="title">Hello & Hi.</div>",
+}
+`
+
+describe('getMatchingLineNumbers', () => {
+  it('gets matched line numbers by exact match', () => {
+    expect(getMatchingLineNumbers(code, 'true', false)).eql([7, 8, 9, 10, 11, 12, 13])
+    expect(getMatchingLineNumbers(code, ' ', false)).eql([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])
+    expect(getMatchingLineNumbers(code, '<', false)).eql([21])
+    expect(getMatchingLineNumbers(code, 'kong', false)).eql([])
+  })
+
+  it('gets matched line numbers by exact match with case sensitivity', () => {
+    expect(getMatchingLineNumbers(code, 'TRUE', false)).eql([])
+  })
+
+  it('gets matched line numbers by regexp match', () => {
+    expect(getMatchingLineNumbers(code, 'tru.', true)).eql([7, 8, 9, 10, 11, 12, 13])
+    expect(getMatchingLineNumbers(code, '[ ]', true)).eql([2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21])
+    expect(getMatchingLineNumbers(code, '.', true)).eql([1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19, 20, 21, 22])
+    expect(getMatchingLineNumbers(code, '<[^>]+>', true)).eql([21])
+    expect(getMatchingLineNumbers(code, 'kong', true)).eql([])
+  })
+
+  it('gets matched line numbers by regexp match with case sensitivity', () => {
+    expect(getMatchingLineNumbers(code, 'TRU.', true)).eql([])
+  })
+})
+
+describe('escapeInnerHTML', () => {
+  it('escapes only < and &', () => {
+    expect(escapeInnerHTML('<script>alert("hi")</script>')).eq('&lt;script>alert("hi")&lt;/script>')
+    expect(escapeInnerHTML('<div>&larr; & &rarr;</div>')).eq('&lt;div>&amp;larr; &amp; &amp;rarr;&lt;/div>')
+    expect(escapeInnerHTML('foo')).eq('foo')
+    expect(escapeInnerHTML('')).eq('')
+  })
+})
+
+describe('escapeHTMLIfNeeded', () => {
+  it('escapes only < and &', () => {
+    const regex = /<&>/
+    const escape = cy.spy((v: string) => v)
+
+    expect(escapeHTMLIfNeeded('{ foo: "bar" }', regex, escape)).eq('{ foo: "bar" }')
+    // eslint-disable-next-line @typescript-eslint/no-unused-expressions
+    expect(escape).not.have.been.called
+  })
+
+  it('escapes only < and & by default', () => {
+    expect(escapeHTMLIfNeeded('<script>alert("hi")</script>')).eq('&lt;script>alert("hi")&lt;/script>')
+    expect(escapeHTMLIfNeeded('<div>&larr; & &rarr;</div>')).eq('&lt;div>&amp;larr; &amp; &amp;rarr;&lt;/div>')
+    expect(escapeHTMLIfNeeded('foo')).eq('foo')
+    expect(escapeHTMLIfNeeded('')).eq('')
+  })
+})
+
+describe('highlightMatchingChars', () => {
+  it('wraps matched characters with <mark class="matched-term"> matched by exact match', () => {
+    expect(highlightMatchingChars(code, 'true', false)).eq(`{
+  "compilerOptions": {
+    "target": "es2020",
+    "module": "esnext",
+    "moduleResolution": "node",
+    "allowUnreachableCode": false,
+    "exactOptionalPropertyTypes": <mark class="matched-term">true</mark>,
```

Comment:
suggestion: can you add the mark HTML into a util function here in the test file so that you can call the function with the expected content and output rather than repeating the HTML throughout all of the tests? 

Ideally you could just import your wrap function from the actual util file in case the actual wrapping HTML changes, it won't break the tests.

Reply:
Because though the major part of the code stays the same, the highlight function modifies the initial code in a non-trivial manner so it's kind of hard to extract a "render function" for this piece of code input.

Reply:
I was just suggesting something like this on the lines that should be highlighted

```ts
import { wrapMark } from './codeBlockHelpers'

// ... test code

"noFallthroughCasesInSwitch": `${wrapMark(true)}`,
```

Reply:
Oh I misread this. That makes sense.